### ISSUE TITLE
torch.sparse_coo_tensor._values() incorrectly returns a detached tensor

### ISSUE THREAD

**Issue by smorad** [pytorch/pytorch#70357]
### 🐛 Describe the bug

Calling `torch.sparse_coo_tensor._values()` incorrectly detaches the tensor from the autograd graph:

```python
import torch
a = torch.sparse_coo_tensor(
  indices=torch.tensor([[1,2,3]]),
  values=torch.tensor([3., 4., 5.], requires_grad=True),
  size=(1000,)
)

a.requires_grad
# True
a.coalesce().values().requires_grad
# True

# When we call without coalescing, gradients are lost
a._values().requires_grad
# False
```

### Versions

```
PyTorch version: 1.10.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 11.4 (arm64)
GCC version: Could not collect
Clang version: 13.0.0 (clang-1300.0.29.30)
CMake version: version 3.21.4
Libc version: N/A

Python version: 3.8.12 | packaged by conda-forge | (default, Oct 12 2021, 21:25:50)  [Clang 11.1.0 ] (64-bit runtime)
Python platform: macOS-11.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.4
[pip3] pytorch-seed-rl==0.0.1
[pip3] torch==1.10.0
[pip3] torch-geometric==2.0.2
[pip3] torch-scatter==2.0.9
[pip3] torch-sparse==0.6.12
[pip3] torch-spline-conv==1.2.1
[pip3] torchtyping==0.1.4
[pip3] torchvision==0.9.0a0
[conda] numpy                     1.21.4           py38hbf7bb01_0    conda-forge
[conda] pytorch                   1.10.0          cpu_py38ha3d347f_0    conda-forge
[conda] pytorch-seed-rl           0.0.1                    pypi_0    pypi
[conda] torch-geometric           2.0.2                    pypi_0    pypi
[conda] torch-scatter             2.0.9                    pypi_0    pypi
[conda] torch-sparse              0.6.12                   pypi_0    pypi
[conda] torch-spline-conv         1.2.1                    pypi_0    pypi
[conda] torchtyping               0.1.4                    pypi_0    pypi
[conda] torchvision               0.9.1           py38h4ddca67_0_cpu    conda-forge
```

cc @nikitaved @pearu @cpuhrsch @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

**Comment by albanD** [comment]
Hi,

Any function that starts with `_` is a private function and is not intended to be used by users.
This one in particular is exposing some internal buffer to the sparse Tensor and so this is expected behavior.

Coalescing and getting the values like you do in the first sample is the right way to get the values from your sparse Tensor.

**Comment by smorad** [comment]
@albanD The docs here currently recommend using `_values()`to get the values of an uncoalesced tensor: https://pytorch.org/docs/stable/sparse.html

<img width="837" alt="Screenshot 2021-12-29 at 17 25 20" src="https://user-images.githubusercontent.com/1743889/147683101-b289f191-30cf-40e0-88e6-43a253fcb316.png">

If I don't care about duplicate indices, calling `.coalesce()` each time I want to access a value/index seems expensive. 

This is quite an insidious bug -- I spent over a week trying to find out why my model wasn't learning. Perhaps printing a warning like `Calling ._values() will return a detached tensor` on the first use of `._values()` could prevent others from making the same mistake.

**Comment by albanD** [comment]
The warning sounds reasonable to me when the user calls `_values()` in python on a Tensor that requires grad.
cc @cpuhrsch what do you think?

**Comment by cpuhrsch** [comment]
It's a private function (leading underscore). We should update this documentation and fix the problem at the root. There seems to be two things going on. 1) Our documentation advises to use a private function. Let's look at _values and _indices vs. values and indices. 2) Our slicing logic calls coalesce and could potentially be made more efficient by supporting uncoalesced inputs. Until then printing a warning is very reasonable. @smorad do you want to send a PR that throws this warning in this situation?

**Comment by pearu** [comment]
> calling .coalesce() each time I want to access a value/index seems expensive.

You don't have to do that. Call `.coalesce()` once after the tensor has been constructed and in subsequent operations always use the coalesced variant of the tensor.

**Comment by cpuhrsch** [comment]
I think the problem is that smorad wants to slice (potentially new) COO Tensors repeatedly and that this will force calls to coalesce. @smorad if this is a performance issue can you post a snippet that shows the problem? As in, if you're using _values to avoid a call to coalesce, where would this call otherwise happen? On top of making the workaround work (better), we can fix the issue that caused you to choose it to begin with.

**Comment by pearu** [comment]
Since the issue is raised in the context of autograd, I wonder, does autograd support workflows with uncoalesced tensors _without_ ever coalescing these? @albanD ?
For instance, indexing an element of an uncoalesced COO tensor should return a COO tensor rather than a scalar value if coalescing would be disabled in the indexing operations.

Btw, COO tensors currently do not support slicing (except along the dense dimensions).

**Comment by smorad** [comment]
@cpuhrsch I can add the warning -- where should I put it? Are there python bindings for `torch.sparse`? And you are correct about my use-case.  E.g., if we want to filter out some elements:
```python
import torch

def filter(x):
  # Let's filter some stuff!
  mask = x.values() > 0.5
  y_idx = x.indices()[:,mask]
  y_val = x.values()[mask]
  # Normally, I would have to coalesce y here, even though I'm guaranteed to not have any duplicate indices
  y = torch.sparse_coo_tensor(y_idx, y_val)
  return y

# Warm the cache
x = torch.sparse_coo_tensor(torch.arange(100000).unsqueeze(0), torch.rand(100000), size=(int(1e8),)).coalesce()
filter(x)

# Let's do some timing
import time
s0 = time.time() 
filter(x) 
s1 = time.time()
print(s1 - s0)

s0 = time.time() 
filter(x).coalesce()
s1 = time.time()
print(s1 - s0)
```

I get `0.0025s` without the extra coalesce call and `0.0051s` with it. 


In my code, I do similar filter operations multiple times per forward pass, which tends to add up.

**Comment by cpuhrsch** [comment]
So I guess another way to help you here is to provide masked_select for sparse Tensors?

**Comment by smorad** [comment]
@cpuhrsch that would work for my particular case, but I think there would be more value to others in allowing access to the gradients without coalescing. However, it sounds like a ton of work if @pearu is correct that autograd is not supported for uncoalesced tensors.

I do think the warning is the highest impact per unit effort.

**Comment by cpuhrsch** [comment]
@smorad you can add the warning [to the cpp code](https://github.com/pytorch/pytorch/blob/29d759948ee7cd2275ea60ebd62c311408e3798b/aten/src/ATen/native/sparse/SparseTensor.cpp#L42-L52). You can warn once [using the TORCH_WARN_ONCE function](https://github.com/pytorch/pytorch/blob/9cf0de509fc519bd8460a999b418f822e6543d33/aten/src/ATen/native/RangeFactories.cpp#L18-L23).

There is a long note that explains the [reasoning behind the various accessors for values and indices and how they relate to autograd](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L5132-L5241). We should probably clean this up....

**Comment by smorad** [comment]
@cpuhrsch another example of unneeded coalescing would be changing the size of a sparse tensor, which AFAIK requires creating a new sparse tensor

```python
x = torch.sparse_coo_tensor(torch.arange(10).unsqueeze(0), torch.rand(10), size=(int(1e8),)).coalesce()
# Whoa, this won't fit in memory!
x.to_dense()
# Let's decrease the size
y = torch.sparse_coo_tensor(x._indices(), x._values(), size=(100,))
# Now we have to coalesce y even though the indices and values have not changed :(
```

**Comment by cpuhrsch** [comment]
@smorad - seems like something we should support via [the select operation](https://pytorch.org/docs/master/generated/torch.select.html#torch.select), if I understand your code correctly.

```
x = torch.sparse_coo_tensor(torch.arange(10).unsqueeze(0), torch.rand(10), size=(int(1e8),)).coalesce()

x = x[:100]
```

We can then optimize select to deal efficiently with uncoalesced inputs, if it shows up as a bottleneck.

**Comment by pearu** [comment]
@smorad:

```
y = torch.sparse_coo_tensor(x.indices(), x.values(), size=(100,))
# Now we have to coalesce y even though the indices and values have not changed :(
```

There is a loophole for this that avoids `coalesce` operation (although for small sizes and already coalesced data, the operation should be cheap):
```
>>> x = torch.sparse_coo_tensor(torch.arange(10).unsqueeze(0), torch.rand(10), size=(int(1e8),)).coalesce()
>>> y = torch.sparse_coo_tensor(x.indices(), x.values(), size=(100,))
>>> y.is_coalesced()
False
>>> y._coalesced_(True)
tensor(indices=tensor([[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]]),
       values=tensor([0.7258, 0.9441, 0.9490, 0.8178, 0.3062, 0.4306, 0.1659,
                      0.5125, 0.2889, 0.2043]),
       size=(100,), nnz=10, layout=torch.sparse_coo)
>>> y.is_coalesced()
True
```
where `_coalesced_` just sets coalesce flag and returns self.

@cpuhrsch I agree, supporting slice indexing on sparse tensors would be a nice feature that would allow users *not* to deal with indices/values explicitly in many situations.

**Comment by smorad** [comment]
`_values()` warning is implemented in #71087

**Comment by smorad** [comment]
Documentation changed are in https://github.com/pytorch/pytorch/pull/71088
